Nhse d32 nhskv.i17

martinsumner opened the pull request:
Changes:

- Make timeouts result in 503. 
- Make global secondary index timeout configurable via riak.conf.
- Make configurable change to encode 2i results using thoas not mochijson2.

martinsumner committed:
Make 2i timeout 503

Avoid issue when an {error, timeout} may be passed as {error, {error, timeout} - resulting in general 500 error

martinsumner committed:
Thoas experiment

martinsumner committed:
Update rebar.config

martinsumner committed:
Test thoas as alternative 2i encode

martinsumner committed:
Make thoas default encode for 2i results

martinsumner committed:
Allow for null key map when not required

martinsumner committed:
Switch aae_fold to use thoas

martinsumner committed:
Merge branch 'nhse-develop' into nhse-d32-nhskv.i17

martinsumner committed:
Don't run long-running eunit test by default

martinsumner commented:
Did a quick experiment taking the JSON library from https://github.com/erlang/otp/pull/8111.  There are some issues, as for obvious reasons this is not designed to be compatible with earlier versions.  With a naive implementation, and some hacks to resolve compatibility, it was slower __in an encode test__ than thoas/mochijson2 (though probably due to the hacks).

I think in this case the performance improvements of `thoas` are worth having now without waiting or back-porting OTP-native JSON.

martinsumner commented:
https://github.com/nhs-riak/riak_test/pull/10

martinsumner commented:
Comparison charts for mochijson2/thoas performance.

Absolute:
<img width="1719" alt="JSONCompare_Absolute" src="https://github.com/nhs-riak/riak_kv/assets/1628897/67413ff6-99de-4c5e-9d5b-6d64ba7724c6">

Relative:
<img width="1752" alt="JSONCompare_Relative" src="https://github.com/nhs-riak/riak_kv/assets/1628897/bb7b2a1e-e297-4ce1-a89e-df29c2b70b15">

ThomasArts commented on src/riak_index.erl:
```diff
@@ -512,8 +512,10 @@ decode_continuation(Bin) ->
 %% We use `infinity' as the default to
 %% match the behavior pre 1.4
 add_timeout_opt(undefined, Opts) ->
```
The above comment is now very misleading, because default is now `0`
It still holds, but now 0 is made into infinity. 
Probably explain this better in the comment

ThomasArts commented on src/riak_kv_clusteraae_fsm.erl:
```diff
@@ -556,50 +556,62 @@ finish(clean, State=#state{from={raw, ReqId, ClientPid}}) ->
 json_encode_results(repair_keys_range, ReplResult) ->
-    R = {struct, [{<<"dispatched_count">>, element(2, ReplResult)}]},
-    mochijson2:encode(R);
+    json_encode_results(repl_keys_range, ReplResult);
```
```suggestion
    json_encode_results(dispatched_count, element(2, ReplResult));
```

I like the original, less indirection, better. When reading code as it is now, I need to understand that it goes to dispatched_count in 2 steps

martinsumner committed:
Update following review

martinsumner commented:
Having fixed my use of the json encode in the OTP PR, this now appears to be notably faster than thoas - https://github.com/nhs-riak/riak_kv/pull/20

On OTP 22:

```
Testing Implementation mochijson2
Result set of 1K in 0ms Result set of 2K in 1ms Result set of 3K in 1ms Result set of 5K in 2ms Result set of 8K in 4ms Total time 10ms
Result set of 13K in 9ms Result set of 21K in 19ms Result set of 34K in 21ms Result set of 55K in 38ms Total time 89ms
Result set of 100K in 91ms Result set of 200K in 191ms Result set of 300K in 230ms Result set of 500K in 358ms Total time 872ms
Result set of 1M in 899ms Result set of 2M in 2082ms Result set of 3M in 2373ms Result set of 5M in 3235ms Total time 8590ms
Result set of 8M in 8948ms Result set of 13M in 8785ms Total time 17733ms


Testing Implementation thoas
Result set of 1K in 0ms Result set of 2K in 0ms Result set of 3K in 1ms Result set of 5K in 2ms Result set of 8K in 3ms Total time 8ms
Result set of 13K in 8ms Result set of 21K in 12ms Result set of 34K in 19ms Result set of 55K in 26ms Total time 67ms
Result set of 100K in 68ms Result set of 200K in 122ms Result set of 300K in 187ms Result set of 500K in 254ms Total time 632ms
Result set of 1M in 828ms Result set of 2M in 1421ms Result set of 3M in 1787ms Result set of 5M in 3083ms Total time 7121ms
Result set of 8M in 7341ms Result set of 13M in 7962ms Total time 15303ms


Testing Implementation otp
Result set of 1K in 0ms Result set of 2K in 0ms Result set of 3K in 0ms Result set of 5K in 1ms Result set of 8K in 2ms Total time 6ms
Result set of 13K in 3ms Result set of 21K in 6ms Result set of 34K in 12ms Result set of 55K in 18ms Total time 41ms
Result set of 100K in 29ms Result set of 200K in 60ms Result set of 300K in 89ms Result set of 500K in 141ms Total time 320ms
Result set of 1M in 546ms Result set of 2M in 608ms Result set of 3M in 877ms Result set of 5M in 1982ms Total time 4015ms
Result set of 8M in 5968ms Result set of 13M in 6372ms Total time 12341ms

```

On OTP 25:

```
Testing Implementation mochijson2
Result set of 1K in 0ms Result set of 2K in 0ms Result set of 3K in 1ms Result set of 5K in 1ms Result set of 8K in 2ms Total time 6ms
Result set of 13K in 5ms Result set of 21K in 7ms Result set of 34K in 12ms Result set of 55K in 15ms Total time 40ms
Result set of 100K in 58ms Result set of 200K in 133ms Result set of 300K in 151ms Result set of 500K in 200ms Total time 544ms
Result set of 1M in 783ms Result set of 2M in 1079ms Result set of 3M in 1372ms Result set of 5M in 2601ms Total time 5837ms
Result set of 8M in 5381ms Result set of 13M in 7838ms Total time 13219ms


Testing Implementation thoas
Result set of 1K in 0ms Result set of 2K in 0ms Result set of 3K in 0ms Result set of 5K in 0ms Result set of 8K in 1ms Total time 4ms
Result set of 13K in 5ms Result set of 21K in 7ms Result set of 34K in 9ms Result set of 55K in 11ms Total time 33ms
Result set of 100K in 44ms Result set of 200K in 75ms Result set of 300K in 92ms Result set of 500K in 126ms Total time 339ms
Result set of 1M in 466ms Result set of 2M in 962ms Result set of 3M in 963ms Result set of 5M in 1237ms Total time 3629ms
Result set of 8M in 5073ms Result set of 13M in 5978ms Total time 11051ms


Testing Implementation otp
Result set of 1K in 0ms Result set of 2K in 0ms Result set of 3K in 0ms Result set of 5K in 1ms Result set of 8K in 1ms Total time 3ms
Result set of 13K in 1ms Result set of 21K in 3ms Result set of 34K in 7ms Result set of 55K in 8ms Total time 20ms
Result set of 100K in 14ms Result set of 200K in 29ms Result set of 300K in 43ms Result set of 500K in 65ms Total time 153ms
Result set of 1M in 390ms Result set of 2M in 579ms Result set of 3M in 454ms Result set of 5M in 1180ms Total time 2604ms
Result set of 8M in 4166ms Result set of 13M in 5133ms Total time 9299ms
```

martinsumner committed:
Encoding use JSON library for OTP PR (#20)

* Encoding use JSON library for OTP PR

Uses the PR'd Json library to OTP for encoding, and passing in a bespoke override on the encode function to avoid the double-pass to convert into a map.

This appears to be notably faster than mochijson2 and thoas.

* Tidy-up

martinsumner commented:
Relative performance at different result sizes - mochijson2 vs OTP Json - OTP 22/24/26:

<img width="1824" alt="JSON Compare - Relative" src="https://github.com/nhs-riak/riak_kv/assets/1628897/aed78154-2e7e-4666-9ec6-fc048119e70c">

martinsumner committed:
Revert thoas addition

Focus on proposed JSON library for OTP.

The riak_kv_clusteraae_fsm has also been reverted back to mochijson2.  The format of leveled_tictac:export_tree isn't compatible with the OTP json library - rather than over-complicate the PR, as performance is not critical for these queries, keep the old implementation.

martinsumner committed:
Test integer index

martinsumner committed:
Extend unit test

martinsumner committed:
Switch to raw OTP PR eaed0f7

Without compatibility changes

martinsumner committed:
Correct filename

martinsumner committed:
Compatibility changes to OTP JSON

Compatibility changes required for OTP 22 - 26

ThomasArts commented on src/riak_kv_wm_json.erl:
```diff
@@ -90,9 +152,9 @@ standards. The decoder is tested using [JSONTestSuite](https://github.com/nst/JS
--doc """
-Simple JSON value encodeable with `json:encode/1`.
-""".
+% -doc """
```
Since you commented this out (possibly because it refers to `json:` you may want to actually remove it?

ThomasArts commented on src/riak_kv_wm_json.erl:
```diff
@@ -18,7 +18,17 @@
 %%
--module(riak_kv_wm_otpjson).
+-module(json).
+-moduledoc """
+A library for encoding and decoding JSON.
+
+This module implements EEP68.
+
+Both encoder and decoder fully conform to
+[RFC 8259](https://tools.ietf.org/html/rfc8259) and
+[ECMA 404](http://www.ecma-international.org/publications/standards/Ecma-404.htm)
+standards. The decoder is tested using [JSONTestSuite](https://github.com/nst/JSONTestSuite).
+""".
```
The triple quotes """ have a new meaning in OTP27 and therefore you get a warning when compiling this code with OTP26... whereas it actually is what you want it won't compile :-(.
 
 Since you want to drop the code in OTP27, you probably want to have it still compile without warnings in OTP26.
 
 ```
%% @doc
%% A library for encoding and decoding JSON.
%%
%% This module implements EEP68.
%%
%% Both encoder and decoder fully conform to
%% [RFC 8259](https://tools.ietf.org/html/rfc8259) and
%% [ECMA 404](http://www.ecma-international.org/publications/standards/Ecma-404.htm)
%% standards. The decoder is tested using [JSONTestSuite](https://github.com/nst/JSONTestSuite).
```

ThomasArts committed:
Update and include quickcheck properties (#21)

* Update and include quickcheck properties

* Output too verbose with a collect.

martinsumner committed:
Update comment as per review
